internal/wasm/interpreter/interpreter.go

```diff
@@ -19,43 +20,63 @@ var callStackCeiling = buildoptions.CallStackCeiling
 
 // engine is an interpreter implementation of internalwasm.Engine
 type engine struct {
-	// Stores compiled functions.
-	functions map[wasm.FunctionAddress]*interpreterFunction
+	// functions stores compiled functions where the index means wasm.FunctionAddress.
+	functions []*interpreterFunction
+
+	// mux is used for read/write access to functions slice.
+	// This is necessary as each compiled function will access the slice at runtime
+	// when they make function calls while engine might be modifying the underlying slice when
+	// adding a new compiled function. We take read lock when creating new virtualMachine
+	// for each function invocation while take write lock in engine.addCompiledFunction.
+	mux sync.RWMutex
+}
+
+const initialCompiledFunctionsSliceSize = 128
+
+func NewEngine() wasm.Engine {
+	return &engine{
+		functions: make([]*interpreterFunction, initialCompiledFunctionsSliceSize),
+	}
+}
+
+// virtualMachine holds context per engine.Call, and shared across all the
```

Comment:
maybe call this rootCall?

Reply:
oh you mean rootCall holds `stack []uint64` and `callFrame []*callFrame` ?

Reply:
I don't think it would be appropriate

Reply:
other runtimes call this kind of struct `vmCtx` so I followed it but I'm open to rename to similar one